feat(server): wait five minutes before sending email on new album item

HeyBanditoz opened the pull request:
Album update jobs will now wait five minutes to send. If a new image is added while that job is pending, the old job will be cancelled, and a new one will be enqueued for a minute.

This is to prevent a flood of notifications by dragging in images directly to the album, which adds them to the album one at a time.

Album updates now include a list of users to email, which is generally everybody except the updater. If somebody else updates the album within that minute, both people will get an album update email in a minute, as they both added images and the other should be notified.

HeyBanditoz commented on server/src/repositories/job.repository.ts:
```diff
@@ -249,24 +251,53 @@ export class JobRepository implements IJobRepository {
+        break;
       }
     }
+    if ((item?.data as IBaseJob)?.delay) {
```
This feels weird, I'm not sure the right TypeScript-y way to do this.

mertalev replied:
Instead of adding `delay` to `IBaseJob` as an optional field , you can make a different `IDelayedJob` that requires it and extend `INotifyAlbumUpdateJob` from it. Then you can move this to the case statement for `JobName.NOTIFY_ALBUM_UPDATE` and hopefully get better typing.

HeyBanditoz replied:
Fixed, it looks much better now, thanks!

mertalev commented on server/src/repositories/job.repository.ts:
```diff
@@ -249,24 +251,53 @@ export class JobRepository implements IJobRepository {
     return this.moduleReference.get<Queue>(getQueueToken(queue), { strict: false });
   }
+
+  public async replaceQueuedOrInsertJob(item: JobItem, dataReplacer: DataTransformer): Promise<void> {
```
You don't need to do any of this. Queueing a job with the same ID as another job will replace it.

HeyBanditoz replied:
I'm not sure how to get that behavior, when I was testing earlier with `IJobRepository#queue` (which calls `#queueAll`) with the same job IDs, it silently drops the job until the first queued job is finished running. I have to get the job from bullmq, then remove it before enqueueing a new one.

mertalev replied:
Oh darn, I thought it replaced the existing job, but it apparently just drops the new one. Removing and re-adding works too I guess. You can simplify it a bit by fetching by job ID instead of getting all the jobs and filtering.

mertalev replied:
I think it'd be better to have a simple `remove` method to remove by a job ID, then maybe an `upsert` method that internally calls `remove` and `queue`.

mertalev replied:
Actually, jobs also have a `changeDelay` method that does what we need in this case.

HeyBanditoz replied:
I also have to change the data, the logic essentially does this:
* if an existing job to notify an album is found,
* remove it,
* check the recipients from the previous job, and add more people to notify if needed.

Say person p1 adds an image, and person p2 adds another image 20 seconds later.
If we solely remove the job and replace it, person p2 will never get a notification, as it was replaced with their notification to p1 that they added an image.

mertalev replied:
Hmm, maybe it'd be better to make this a bit lazier. What if you scheduled a delayed job that, when, run, checks the `updatedAt` field of the album? If it's within a certain duration, it can re-queue itself and check again later. You don't need to worry about removing and re-adding this job; you just need to make sure it's queued.

HeyBanditoz replied:
I'm imaging something like this:
* schedule only album ID and timestamp of first update.
* if job is still scheduled, push it back another few minutes but keep time of first update,
* when job fires, get all images added between timestamp on the job and now,
  * and notify those users.

does that sound good? also cleaner since there is no need to keep the users to notify as state, can just calculate that near the instant the notification is sent

mertalev replied:
The first and third bullet points sound good to me.

For the second, I think it'd be better to always try to queue the job and check within the job if it should re-queue itself. The advantages are that it involves fewer requests and is more parallel-safe: when you check if the job exists and create it or change the delay based on that, you're doing 2 requests and conditions can change between those two.

If you queue the job and move on, it's one request with no race condition, taking advantage of the fact that later jobs with the same job ID are dropped (i.e. the timestamp of the first update is preserved).

HeyBanditoz replied:
I think I might have something sorta stubbed out that might work, but I've run into a blocker:

`immich=# SELECT * FROM albums_assets_assets;`
               albumsId               |               assetsId
--------------------------------------|--------------------------------------
 3c8f611f-4d15-4ec6-89bf-343fb3b30156 | 3619bdcf-d8f3-41ff-a5fe-8750ff440e34
 3c8f611f-4d15-4ec6-89bf-343fb3b30156 | d0d3f9e1-24be-49e8-b12d-890d914bee4c

There doesn't seem to be anything that tracks when an image was added to an album, since I'm just passing when the album was updated, I need to get all new images. Would it make sense to add a `createdOn` to this table here? It doesn't look straightforward to accomplish with [typeorm](https://orkhan.gitbook.io/typeorm/docs/many-to-many-relations)

HeyBanditoz replied:
Alright, I removed the weird DataReplacer function param and created a `remove` function on the JobRepository that just returns the job data if it's available to build out the new notification recipient list. Much easier to write a unit test for this case too.

alextran1502 commented on server/src/interfaces/event.interface.ts:
```diff
@@ -14,7 +14,7 @@ type EmitEventMap = {
 
   // album events
-  'album.update': [{ id: string; updatedBy: string }];
+  'album.update': [{ id: string; deliverTo: string[] }];
```
We have been using the recipient term, so maybe here we can use `receipientIds`

HeyBanditoz replied:
Renamed to `recipientIds`.

HeyBanditoz committed:
feat(server): wait five minutes before sending email on new album item

Album update jobs will now wait five minutes to send. If a new image is added while that job is pending, the old job will be cancelled, and a new one will be enqueued for a minute.

This is to prevent a flood of notifications by dragging in images directly to the album, which adds them to the album one at a time.

Album updates now include a list of users to email, which is generally everybody except the updater. If somebody else updates the album within that minute, both people will get an album update email in a minute, as they both added images and the other should be notified.

alextran1502 commented:
@mertalev @jrasm91 anything you guys would like to address in this PR?

mertalev commented:
We talked about different approaches for notifications, but came to the conclusion that this approach is fine for now until we have a more detailed plan on how notifications should be scheduled.

HeyBanditoz commented:
@mertalev @alextran1502 I've resolved conflicts, this should be ready for review now

mblissmer commented:
Just a thought: i'd love to see this either configurable (which would be a lot more work I assume), or have 60 seconds bumped up to something way higher, like 15 minutes or something. Like, why not?

simono41 commented:
> Just a thought: i'd love to see this either configurable (which would be a lot more work I assume), or have 60 seconds bumped up to something way higher, like 15 minutes or something. Like, why not?

Let's get this pull request through first. I just went on vacation with some friends and they always complain when someone uploads one or more pictures. I had to temporarily completely deactivate the email notification function until further notice so that your inbox wouldn't be flooded.

HeyBanditoz commented:
Yeah I picked 60 seconds since it takes a quarter of that time to upload an image one-by-one to my VPS running Immich.

If I recall correctly. a workaround to prevent the notification flood is to upload all the images to your user, then navigate to the album and bulk add them there (plus at top right on web) so they aren't added one-by-one.

mertalev commented on server/src/services/notification.service.ts:
```diff
@@ -100,8 +104,30 @@ export class NotificationService extends BaseService {
-  async onAlbumUpdate({ id, updatedBy }: ArgOf<'album.update'>) {
-    await this.jobRepository.queue({ name: JobName.NOTIFY_ALBUM_UPDATE, data: { id, senderId: updatedBy } });
+  async onAlbumUpdate({ id, recipientIds }: ArgOf<'album.update'>) {
+    // if deliverTo is empty, album likely only has one user part of it, don't queue notification if so
```
`deliverTo` is outdated

mertalev commented on server/src/repositories/job.repository.ts:
```diff
@@ -249,24 +250,42 @@ export class JobRepository implements IJobRepository {
+    if (existingJob === undefined) {
+      return undefined;
+    }
+    try {
+      await existingJob.remove();
+    } catch {
+      return undefined;
+    }
```
This is sketchy. Best to check the error to make sure it's what you expect (which I assume is that the job doesn't exist)

HeyBanditoz replied:
Yeah in this case I didn't want job doesn't exist errors to bubble up, which I think would be rare since we just got the job a few lines above, so I just returned `undefined` instead, are you saying I should catch the specific job not found exception, but let other errors bubble up?

mertalev replied:
Yes, something like a connection error etc. shouldn't be ignored.

HeyBanditoz replied:
I'm not sure how to handle this one, it look like bullmq internally only throws `Error`s, https://github.com/taskforcesh/bullmq/blob/9307fef3e5f9e5f929fd8755078913a33c2dc783/src/classes/job.ts#L601-L603 and https://github.com/taskforcesh/bullmq/blob/9307fef3e5f9e5f929fd8755078913a33c2dc783/src/classes/scripts.ts#L1423

mertalev replied:
Oh, this is the fun part about JS. *Everyone* just throws an `Error` so you have to check the message etc. to know what kind of error it is.

HeyBanditoz replied:
Would you check it by making a new error type or just checking if the error string contains `Missing key for job`?

mertalev replied:
String matching is fine

mertalev commented on server/src/services/notification.service.ts:
```diff
@@ -100,8 +104,30 @@ export class NotificationService extends BaseService {
+
+    const previousJobData = await this.jobRepository.removeJob(id, JobName.NOTIFY_ALBUM_UPDATE);
+    if (previousJobData !== undefined && this.isAlbumUpdateJob(previousJobData)) {
+      for (const string of previousJobData.recipientIds) {
```
`id` would be better than `string`

mertalev commented on server/src/services/notification.service.ts:
```diff
@@ -100,8 +104,30 @@ export class NotificationService extends BaseService {
+    };
+
+    const previousJobData = await this.jobRepository.removeJob(id, JobName.NOTIFY_ALBUM_UPDATE);
+    if (previousJobData !== undefined && this.isAlbumUpdateJob(previousJobData)) {
```
You can just check if it's truthy in the first clause

mertalev commented on server/src/repositories/job.repository.ts:
```diff
@@ -249,24 +250,42 @@ export class JobRepository implements IJobRepository {
+
+  public async removeJob(jobId: string, name: JobName): Promise<IEntityJob | undefined> {
+    const existingJob = await this.getQueue(JOBS_TO_QUEUE[name]).getJob(jobId);
+    if (existingJob === undefined) {
```
You can just check if it's falsy

mertalev commented on server/src/services/notification.service.ts:
```diff
@@ -100,8 +104,30 @@ export class NotificationService extends BaseService {
-    await this.jobRepository.queue({ name: JobName.NOTIFY_ALBUM_UPDATE, data: { id, senderId: updatedBy } });
+  async onAlbumUpdate({ id, recipientIds }: ArgOf<'album.update'>) {
+    // if deliverTo is empty, album likely only has one user part of it, don't queue notification if so
+    if (!recipientIds) {
```
Arrays are truthy, even empty ones.

HeyBanditoz replied:
oh now IDE is saying `recipientIds` being `undefined` is impossible, I'll just keep the length check only, am too used to Java world trusting no one 🙃

mertalev commented:
I think 60s is a little too short considering videos and slower connections. 300s would be nicer imo

mertalev commented on server/src/services/album.service.ts:
```diff
@@ -174,7 +174,11 @@ export class AlbumService extends BaseService {
+        (userId) => userId !== auth.user.id,
+      );
+
+      await this.eventRepository.emit('album.update', { id, recipientIds: allUsersExceptUs });
```
You can check if `allUsersExceptUs` is empty before emitting in the first place.

mertalev commented on server/src/repositories/job.repository.ts:
```diff
@@ -249,24 +250,46 @@ export class JobRepository implements IJobRepository {
+    } catch (error: any) {
+      if (error.message?.includes('Missing key for job')) {
+        return undefined;
+      } else {
+        throw error;
+      }
```
An `else` block is unnecessary here

HeyBanditoz replied:
but wouldn't the catch also swallow the other exception types without it?

mertalev replied:
No, I mean that you can just put `throw error;` there because the only way to reach that line is if the `if` statement didn't apply.

HeyBanditoz replied:
Oh 🤦 tyty

danieldietzler commented on server/src/repositories/job.repository.ts:
```diff
@@ -249,24 +250,45 @@ export class JobRepository implements IJobRepository {
+        opts = { jobId: item.data.id, delay: item.data?.delay };
+        break;
+      }
       case JobName.STORAGE_TEMPLATE_MIGRATION_SINGLE: {
-        return { jobId: item.data.id };
+        opts = { jobId: item.data.id };
+        break;
       }
       case JobName.GENERATE_PERSON_THUMBNAIL: {
-        return { priority: 1 };
+        opts = { priority: 1 };
+        break;
       }
       case JobName.QUEUE_FACIAL_RECOGNITION: {
-        return { jobId: JobName.QUEUE_FACIAL_RECOGNITION };
-      }
-
-      default: {
-        return null;
+        opts = { jobId: JobName.QUEUE_FACIAL_RECOGNITION };
+        break;
       }
     }
+    return opts;
   }
```
Why did you add the `opts` variable? Just return in every case.

HeyBanditoz replied:
This was a leftover from a previous impl, nuked it

danieldietzler commented on server/src/repositories/job.repository.ts:
```diff
@@ -249,24 +250,45 @@ export class JobRepository implements IJobRepository {
+  public async removeJob(jobId: string, name: JobName): Promise<IEntityJob | undefined> {
+    const existingJob = await this.getQueue(JOBS_TO_QUEUE[name]).getJob(jobId);
+    if (!existingJob) {
+      return undefined;
```
```suggestion
      return;
```

danieldietzler commented on server/src/repositories/job.repository.ts:
```diff
@@ -249,24 +250,45 @@ export class JobRepository implements IJobRepository {
+      await existingJob.remove();
+    } catch (error: any) {
+      if (error.message?.includes('Missing key for job')) {
+        return undefined;
```
```suggestion
        return;
```

danieldietzler commented on server/src/services/album.service.ts:
```diff
@@ -174,7 +174,13 @@ export class AlbumService extends BaseService {
       });
 
-      await this.eventRepository.emit('album.update', { id, updatedBy: auth.user.id });
+      const allUsersExceptUs = [...album.albumUsers.map((au) => au.user.id), album.owner.id].filter(
```
Please don't use abbreviations. (such as `au`)

danieldietzler replied:
In fact, you can use destructuring and just do `({ user }) => user.id`

mertalev commented on server/src/interfaces/job.interface.ts:
```diff
@@ -293,6 +298,7 @@ export enum JobStatus {
 
 export type JobHandler<T = any> = (data: T) => Promise<JobStatus>;
 export type JobItemHandler = (item: JobItem) => Promise<void>;
+export type DataTransformer = (fromExistingJob: any, fromEnqueueingJob: any) => any;
```
Is this used anywhere?

HeyBanditoz replied:
nope, it was in a previous impl, nuked it

danieldietzler commented on server/src/services/notification.service.spec.ts:
```diff
@@ -607,11 +590,25 @@ describe(NotificationService.name, () => {
     });
+
+    it('should add new recipients for new images if job is already queued', async () => {
+      // @ts-expect-error needed to force a INotifyAlbumUpdateJob
```
You should just `as INotifyAlbumUpdateJob` cast it.
